### PR TITLE
Implement retries (#56)

### PR DESCRIPTION
**Pull request by Spiritreader**
## Overview

This pull request implements the retry functionality mentioned in #21

It follows the outlined draft (#56) pretty closely, except for a bit more frontend fixes to display the correct color and strings when the service is pending (retrying)

Also for now, if a service enters the retry condition, hearbeats with that status do not count towards the uptime calculation.

### This change requires a database migration for existing uptime kuma installations.

The new table looks like this:
```
CREATE TABLE "monitor" (
	"id"	INTEGER NOT NULL,
	"name"	VARCHAR(150),
	"active"	BOOLEAN NOT NULL DEFAULT 1,
	"user_id"	INTEGER,
	"interval"	INTEGER NOT NULL DEFAULT 20,
	"url"	TEXT,
	"type"	VARCHAR(20),
	"weight"	INTEGER DEFAULT 2000,
	"hostname"	VARCHAR(255),
	"port"	INTEGER,
	"created_date"	TIMESTAMP,
	"keyword"	VARCHAR(255),
	"maxretries"	INTEGER NOT NULL DEFAULT 0,
	PRIMARY KEY("id" AUTOINCREMENT),
	FOREIGN KEY("user_id") REFERENCES "user"("id") ON DELETE SET NULL ON UPDATE CASCADE
);
```

The monitor table received another field `maxretries`

The required migration script is **NOT** part of this PR, however the **initial database has been updated accordingly.**

## Screenshot
image

### CODE REVIEW

File: src/pages/EditMonitor.vue [hunk 1]
```diff
@@ -48,6 +48,11 @@
                         <input type="number" class="form-control" id="interval" v-model="monitor.interval" required min="20" step="1">
                     </div>
 
+                      <div class="mb-3">
+                        <label for="interval" class="form-label">Retry Pings</label>
+                        <input type="number" class="form-control" id="interval" v-model="monitor.maxretries" required min="0" step="1">
+                    </div>
```
Comment (Saibamen) [inline src/pages/EditMonitor.vue:54]: Fix for indentation
```suggestion
                    <div class="mb-3">
                        <label for="interval" class="form-label">Retry Pings</label>
                        <input type="number" class="form-control" id="interval" v-model="monitor.maxretries" required min="0" step="1">
                    </div>
```

File: .gitignore [hunk 2]
```diff
@@ -7,3 +7,4 @@ dist-ssr
 
 /data
 !/data/.gitkeep
+.vscode
```
Comment (Saibamen) [inline .gitignore:10]: Please also add it to `.dockerignore` file

**Review by Saibamen — CHANGES_REQUESTED** [review]

File: .gitignore (hunk 2 above)
Comment (Spiritreader) [reply .gitignore:10]: done, thanks

**Review by Saibamen — APPROVED** [review]

**Comment by Saibamen** [comment]
@louislam: Do we have automatic DB migrations at server startup?

**Comment by louislam** [comment]
> @louislam: Do we have automatic DB migrations at server startup?

Yes, I am implementing this.

File: src/pages/EditMonitor.vue [hunk 3]
```diff
@@ -48,6 +48,11 @@
                         <input type="number" class="form-control" id="interval" v-model="monitor.interval" required min="20" step="1">
                     </div>
 
+                    <div class="mb-3">
+                        <label for="maxRetries" class="form-label">Retry Pings</label>
```
Comment (louislam) [inline src/pages/EditMonitor.vue:52]: Looing for better name for this, "Retry Pings" looks strange to me.

Comment (Saibamen) [reply src/pages/EditMonitor.vue:52]: Just `Retries`?
And IMO will be nice to also add some "tip" (outside `<label>`) like: "Maximum retries before sending a notification"

Comment (Spiritreader) [reply src/pages/EditMonitor.vue:52]: Ye I'm not very good at naming things.
Retries sounds a lot better.

As for the hint, how about "Maximum retries before service is marked as down"?

Comment (Saibamen) [reply src/pages/EditMonitor.vue:52]: LGTM

File: src/pages/EditMonitor.vue [hunk 4]
```diff
@@ -49,8 +49,9 @@
                     </div>
 
                     <div class="mb-3">
-                        <label for="maxRetries" class="form-label">Retry Pings</label>
+                        <label for="maxRetries" class="form-label">Retries</label>
                         <input type="number" class="form-control" id="maxRetries" v-model="monitor.maxretries" required min="0" step="1">
+                        <div class="form-text">Maximum retries before the service is marked as down and send notifications</div>
```
Comment (Spiritreader) [inline src/pages/EditMonitor.vue:54]: I believe
`"Maximum retries before the service is marked as down and a notification is sent"`
would be more grammatically correct.
Any thoughts?

**Pull request merged**